File: src/verifier.ts
import { Point, AlignmentResult, RebuttalSchema } from './types';

/**
 * A flat list of (point, factCitation) pairs ready to be sent to the
 * Source-Alignment Verifier. Statute citations are excluded — those are governed
 * by the Citation Gate's existence check; the Verifier audits FACT alignment only.
 */
export interface VerifierTask {
  pointIndex: number;
  pointSource: string;
  claim: string;
  citationId: string;
}

const FACT_ID = /^F\d+/i;

/**
 * Walks every cited claim in the debate transcript and produces one VerifierTask
 * per (point, factId) pair. Skips statute citations.
 */
export function collectVerifierTasks(args: {
  advocatePoints: Point[];
  opposingTheory: Point[];
  attackPoints: Point[];
  rebuttal: ReturnType<typeof RebuttalSchema.parse>;
}): VerifierTask[] {
  const tasks: VerifierTask[] = [];
  pushAll(args.advocatePoints, 'advocate_position', tasks);
  pushAll(args.opposingTheory, 'opposing_independent', tasks);
  pushAll(args.attackPoints, 'opposing_attack', tasks);
  // Rebuttal items have the same {claim, citations} shape, plus a stance.
  args.rebuttal.responses.forEach((r, i) => {
    for (const c of r.citations) {
      if (FACT_ID.test(c)) {
        tasks.push({
          pointIndex: i,
          pointSource: `advocate_rebuttal:${r.stance}`,
          claim: r.claim,
          citationId: c,
        });
      }
    }
  });
  return tasks;
}

function pushAll(points: Point[], source: string, out: VerifierTask[]): void {
  points.forEach((p, i) => {
    for (const c of p.citations) {
      if (FACT_ID.test(c)) {
        out.push({ pointIndex: i, pointSource: source, claim: p.claim, citationId: c });
      }
    }
  });
}

export interface VerifierSummary {
  total: number;
  supported: number;
  contradicted: number;
  overreach: number;
  neutral: number;
  contradictedDetails: AlignmentResult[];
  overreachDetails: AlignmentResult[];
}

export function summarizeAlignment(results: AlignmentResult[]): VerifierSummary {
  const summary: VerifierSummary = {
    total: results.length,
    supported: 0,
    contradicted: 0,
    overreach: 0,
    neutral: 0,
    contradictedDetails: [],
    overreachDetails: [],
  };
  for (const r of results) {
    if (r.alignment === 'supported') summary.supported++;
    else if (r.alignment === 'contradicted') {
      summary.contradicted++;
      summary.contradictedDetails.push(r);
    } else if (r.alignment === 'overreach') {
      summary.overreach++;
      summary.overreachDetails.push(r);
    } else summary.neutral++;
  }
  return summary;
}

File: lumen_py/verifier.py
"""Source-Alignment Verifier helpers (ported from src/verifier.py).

Collects (claim, factId) pairs from the transcript and summarizes the verifier's
alignment verdicts. Statute citations are excluded — those are the Citation Gate's job.
"""
from __future__ import annotations
import re
from dataclasses import dataclass, field

from .types import Point, Rebuttal, AlignmentResult

_FACT_ID = re.compile(r"^F\d+", re.IGNORECASE)


@dataclass
class VerifierTask:
    pointIndex: int
    pointSource: str
    claim: str
    citationId: str


def collect_verifier_tasks(
    advocate_points: list[Point],
    opposing_theory: list[Point],
    attack_points: list[Point],
    rebuttal: Rebuttal,
) -> list[VerifierTask]:
    tasks: list[VerifierTask] = []

    def push_all(points: list[Point], source: str) -> None:
        for i, p in enumerate(points):
            for c in p.citations:
                if _FACT_ID.match(c):
                    tasks.append(VerifierTask(i, source, p.claim, c))

    push_all(advocate_points, "advocate_position")
    push_all(opposing_theory, "opposing_independent")
    push_all(attack_points, "opposing_attack")
    for i, r in enumerate(rebuttal.responses):
        for c in r.citations:
            if _FACT_ID.match(c):
                tasks.append(VerifierTask(i, f"advocate_rebuttal:{r.stance}", r.claim, c))
    return tasks


@dataclass
class VerifierSummary:
    total: int = 0
    supported: int = 0
    contradicted: int = 0
    overreach: int = 0
    neutral: int = 0
    contradicted_details: list[AlignmentResult] = field(default_factory=list)
    overreach_details: list[AlignmentResult] = field(default_factory=list)


def summarize_alignment(results: list[AlignmentResult]) -> VerifierSummary:
    s = VerifierSummary(total=len(results))
    for r in results:
        if r.alignment == "supported":
            s.supported += 1
        elif r.alignment == "contradicted":
            s.contradicted += 1
            s.contradicted_details.append(r)
        elif r.alignment == "overreach":
            s.overreach += 1
            s.overreach_details.append(r)
        else:
            s.neutral += 1
    return s
